avro/src/main/scala/hydra/avro/registry/SchemaRegistry.scala

```diff
@@ -64,38 +66,65 @@ trait SchemaRegistry[F[_]] {
     */
   def getSchemaRegistryClient: F[SchemaRegistryClient]
 
-  def getSchemaBySubject(subject: String): F[Option[Schema]]
+  /**
+    * Retrieves the latest schema for the given subject name, if exists
+    * @param subject - subject name for the schema found in SchemaRegistry including the suffix (-key | -value)
+    * @return - Optional Schema for the given subject name
+    */
+  def getLatestSchemaBySubject(subject: String): F[Option[Schema]]
+
+  /**
+    * Retrieves schema for the version and subject specified, if exists
+    * @param subject - subject name for the schema found in SchemaRegistry including the suffix (-key | -value)
+    * @param schemaVersion - version number for the schema
+    * @return Optional Schema for the given subject and version combination
+    */
+  def getSchemaFor(subject: String, schemaVersion: SchemaVersion): F[Option[Schema]]
 
 }
 
 object SchemaRegistry {
 
+  final case class IncompatibleSchemaException(message: String) extends
+    RuntimeException(message)
+
   type SchemaId = Int
   type SchemaVersion = Int
 
+
+  private[registry] def validate(newSchema: Schema, oldSchemas: List[Schema]): Boolean = {
+    AvroCompatibilityChecker.FULL_TRANSITIVE_CHECKER.isCompatible(newSchema, oldSchemas.asJava)
+  }
+
   def live[F[_]: Sync](
       schemaRegistryBaseUrl: String,
       maxCacheSize: Int
   ): F[SchemaRegistry[F]] = Sync[F].delay {
-    getFromSchemaRegistryClient(
-      new CachedSchemaRegistryClient(schemaRegistryBaseUrl, maxCacheSize)
-    )
+    getFromSchemaRegistryClient(new CachedSchemaRegistryClient(schemaRegistryBaseUrl, maxCacheSize))
   }
 
   def test[F[_]: Sync]: F[SchemaRegistry[F]] = Sync[F].delay {
     getFromSchemaRegistryClient(new MockSchemaRegistryClient)
   }
 
-  private[this] def getFromSchemaRegistryClient[F[_]: Sync](
-      schemaRegistryClient: SchemaRegistryClient
-  ): SchemaRegistry[F] =
+  private def getFromSchemaRegistryClient[F[_]: Sync](schemaRegistryClient: SchemaRegistryClient): SchemaRegistry[F] =
     new SchemaRegistry[F] {
 
       override def registerSchema(
           subject: String,
           schema: Schema
-      ): F[SchemaId] =
-        Sync[F].delay(schemaRegistryClient.register(subject, schema))
+      ): F[SchemaId] = {
+        for {
+          versions <- getAllVersions(subject)
+          schemas <- versions.traverse(getSchemaFor(subject, _)).map(_.flatten)
+          validated <- Sync[F].delay(validate(schema, schemas.reverse))
+          schemaVersion <- if (validated) {
+            Sync[F].delay(schemaRegistryClient.register(subject, schema))
+          } else {
+            Sync[F].raiseError[SchemaVersion](IncompatibleSchemaException("Incompatible Schema Evolution. You may add fields with default fields, or remove fields with default fields."))
+          }
+        } yield schemaVersion
```

Comment:
Is this function ever tested? It is possible that it was covered by existing tests, but I just want to make sure we are testing the scenario:
1. Add schema A
2. Add schema A with evolution

Reply:
Ooo, good question. I'll check it out.